test/ziggurat/init_test.clj

```diff
@@ -3,26 +3,36 @@
             [mount.core :refer [defstate] :as mount]
             [ziggurat.config :as config]
             [ziggurat.init :as init]
-            [ziggurat.messaging.rabbitmq-wrapper :as rmqw]
             [ziggurat.messaging.consumer :as messaging-consumer]
             [ziggurat.messaging.producer :as messaging-producer]
             [ziggurat.streams :as streams :refer [stream]]
             [ziggurat.server.test-utils :as tu]
-            [ziggurat.tracer :as tracer])
+            [ziggurat.tracer :as tracer]
+            [ziggurat.messaging.messaging :as messaging])
   (:import (io.opentracing.mock MockTracer)))
 
+(def valid-modes-count 4)
+
+(defn- exp [x n]
+  (loop [acc 1 n n]
+    (if (zero? n) acc
+                  (recur (* x acc) (dec n)))))
+
 (deftest start-calls-actor-start-fn-test
   (testing "The actor start fn starts before the ziggurat state and can read config"
-    (let [result (atom 1)]
-      (with-redefs [streams/start-streams (fn [_ _] (reset! result (* @result 2)))
-                    streams/stop-streams  (constantly nil)
-                    rmqw/start-connection (fn [_ _] (reset! result (* @result 2)))
-                    rmqw/stop-connection  (constantly nil)
-                    config/config-file    "config.test.edn"
-                    tracer/create-tracer  (fn [] (MockTracer.))]
+    (let [result                              (atom 1)
+          start-messaging-internal-call-count 2]
+      (with-redefs [streams/start-streams      (fn [_ _] (reset! result (* @result 2)))
+                    streams/stop-streams       (constantly nil)
+                    ;; will be called valid modes number of times
+                    messaging/start-connection (fn [_ _] (reset! result (* @result 2)))
+                    messaging/stop-connection  (constantly nil)
+                    config/config-file         "config.test.edn"
+                    tracer/create-tracer       (fn [] (MockTracer.))]
         (init/start #(reset! result (+ @result 3)) {} [] nil)
         (init/stop #() nil)
-        (is (= 16 @result))))))
+        ;; some of the functions which call start-messaging, are called again.
+        (is (= (* 4 (exp 2 (+ start-messaging-internal-call-count valid-modes-count))) @result))))))
```

Comment:
is this due to a change in logic?

Reply:
yes, since we have moved away from defstate. The start-connection will be called as many times as the number of valid modes

Reply:
ohh okay